src/Storages/StorageURLCluster.cpp

```diff
@@ -0,0 +1,159 @@
+#include "config.h"
+#include "Interpreters/Context_fwd.h"
+
+#include <Storages/StorageURLCluster.h>
+
+#include <Client/Connection.h>
+#include <Core/QueryProcessingStage.h>
+#include <DataTypes/DataTypeString.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/getHeaderForProcessingStage.h>
+#include <Interpreters/SelectQueryOptions.h>
+#include <Interpreters/InterpreterSelectQuery.h>
+#include <QueryPipeline/narrowPipe.h>
+#include <QueryPipeline/Pipe.h>
+#include <QueryPipeline/RemoteQueryExecutor.h>
+
+#include <Processors/Transforms/AddingDefaultsTransform.h>
+
+#include <Processors/Sources/RemoteSource.h>
+#include <Parsers/queryToString.h>
+#include <Parsers/ASTTablesInSelectQuery.h>
+
+#include <Storages/IStorage.h>
+#include <Storages/StorageURL.h>
+#include <Storages/SelectQueryInfo.h>
+#include <Storages/StorageDictionary.h>
+#include <Storages/addColumnsStructureToQueryWithClusterEngine.h>
+
+#include <memory>
+
+
+namespace DB
+{
+
+StorageURLCluster::StorageURLCluster(
+    ContextPtr context_,
+    String cluster_name_,
+    const String & uri_,
+    const StorageID & table_id_,
+    const String & format_name_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    const String & compression_method_,
+    const StorageURL::Configuration &configuration_)
+    : IStorageCluster(table_id_)
+    , cluster_name(cluster_name_)
+    , uri(uri_)
+    , format_name(format_name_)
+    , compression_method(compression_method_)
+{
+    context_->getRemoteHostFilter().checkURL(Poco::URI(uri_));
+
+    StorageInMemoryMetadata storage_metadata;
+
+    if (columns_.empty())
+    {
+        auto columns = StorageURL::getTableStructureFromData(format_name_,
+            uri_,
+            chooseCompressionMethod(Poco::URI(uri_).getPath(), compression_method),
+            configuration_.headers,
+            std::nullopt,
+            context_);
+        storage_metadata.setColumns(columns);
+        add_columns_structure_to_query = true;
```

Comment:
I see you almost copied it from `StorageS3Cluster`. It's ok, but please take into account recent fix https://github.com/ClickHouse/ClickHouse/pull/45422

Reply:
I will fix it recently, actually I copied code from `StorageHDFSCluster`, so maybe the same bug in StorageHDFSCluter

Reply:
Yes, there is the same bug there. I will fix it. Thanks for it :)

Reply:
I have doubts about the compatibility of urlCluster and RemoteQueryExecutor, in RemoteQueryExecutor, the TaskIterator only support String
```cpp
using TaskIterator = std::function<String()>;
```
The input of the S3 and HDFS' `initialize` function is a String, so they are compatible.

```cpp
StorageS3Source::ReaderHolder StorageS3Source::createReader()
{
    auto [current_key, info] = (*file_iterator)(); //current_key is a String
    if (current_key.empty())
        return {};
   ....
}
```
```cpp
bool HDFSSource::initialize()
{
    current_path = (*file_iterator)(); //current_path is a String
    if (current_path.empty())
        return false;
    ....
}
```

However, the input of the URL `initialize` function is `vector<String>`.

```cpp
    initialize = [=, this](const URIInfo::FailoverOptions & uri_options) //FailoverOptions is a std::vector<String> 
    {
        if (uri_options.empty())
            throw Exception(ErrorCodes::LOGICAL_ERROR, "Got empty url list");

        auto first_option = uri_options.begin();
    }
```

So the question is how can I pass a std::vector<String> parameter by TaskIterator?

I think there are two ways

1. provide another TaskIterator with type `std::vector<String>`
2. use some trick method to transfer std::vector<String> to String and split it in the cluster.

Reply:
The first one is better.